### ISSUE TITLE
DLS_DEAD_LOCAL_STORE false-positive in switch

### ISSUE THREAD

**Issue by boris-petrov** [spotbugs/spotbugs#2736]
```java
switch (x) {
    case Map<?, ?> m -> return 1;
    default -> return 2;
}
```

This leads to a `DLS_DEAD_LOCAL_STORE` error for the `m` variable. That's a false-positive because Java doesn't allow the variable name to be omitted.

**Comment by gtoison** [comment]
This might be solved if this JEP 456: "Unnamed Variables & Patterns" is implemented: https://openjdk.org/jeps/456

**Comment by boris-petrov** [comment]
@gtoison yes, that's true, once Java 22 is available, the current behavior will be correct. But until then it isn't. :smile:

**Comment by gtoison** [comment]
> @gtoison yes, that's true, once Java 22 is available, the current behavior will be correct. But until then it isn't. 😄

... and once everybody has switched to Java 22  😄
It also remains to be seen how SpotBugs will analyse the bytecode produced with unamed variables

**Comment by boris-petrov** [comment]
I tried SpotBugs `4.8.4` which I thought contains the fix for this but it didn't fix any of my cases. Am I missing something or is this still not released?

**Comment by gtoison** [comment]
Hum, I thought this was fixed by #2828 ...
The test cases in https://github.com/gtoison/spotbugs/blob/b5893ac7a825413dd6085135318ff7a5a7c49e49/spotbugsTestCases/src/java21/Issue2782.java seem to be handled correctly so maybe your code is a bit different?
What compiler (and version) are you using? I suppose that different compiler might be outputting different bytecode here.

The change in #2828 looks for a CHECKCAST followed by an ASTORE at the beginning of a switch case

**Comment by boris-petrov** [comment]
```
% java --version
openjdk 21.0.2 2024-01-16
OpenJDK Runtime Environment (build 21.0.2+13)
OpenJDK 64-Bit Server VM (build 21.0.2+13, mixed mode, sharing)
```

As for the code that doesn't work - as I said, none of my cases (around 10 or so) have been fixed... but the original one I've given here seems to work which is strange. :smile: Here are a couple more:

```java
return switch (value) {
	case Map<?, ?> map -> map.size();
	case Set<?> set -> set.size();
	case List<?> list -> list.size();
	case Collection<?> collection ->
		throw new UnsupportedOperationException("Error " + value.getClass());
	case null, default -> 0;
};
```

Gives the warning on `collection`.

```java
private static String getValueType(Object value) {
	return switch (value) {
		case List<?> list -> "list";
		case Map<?, ?> map -> "map";
		case Number number -> "number";
		case null, default -> "other";
	};
}
```

Gives the warning on all `list`, `map` and `number`.

**Comment by gtoison** [comment]
Thanks for the samples, I have added them in #2937 and they are flagged indeed
I'd have to investigate why the they are not when running the test on my computer

**Comment by maxl2287** [comment]
Updating on 4.8.5.0 is removing this DLS for such switches.

**Comment by gtoison** [comment]
You might run into false positives with 4.8.5 because I had not realized that the bytecode instruction for a switch is either `LOOKUPSWITCH` or `TABLESWITCH`. Your compiler might use `TABLESWITCH` in some cases.
In 4.8.5 only `LOOKUPSWITCH` is handled, once released the fix in #2937 should handle both.

**Comment by boris-petrov** [comment]
I just updated my code-base to use Java 22 and actually this bug is now a great feature! :) Not sure what's the right solution... On Java 22+ it's a feature, on Java 21- is a bug... what are you going to do in the end?

**Comment by gtoison** [comment]
Just to clarify what you're saying: do you mean that you started using JEP 456's "unnamed variables" with Java 22 and you're not getting false positives (because TABLESWITCH switches weren't handled)?
I suppose that when the fix from #2937 gets released you will indeed get false positives

**Comment by boris-petrov** [comment]
No, sorry, I meant that I started using the unnamed variables feature and now I would like to get a warning on the code in my first post here. That is, that code *should* warn in JDK 22+ and should *not* warn on JDK 21-.

**Comment by gtoison** [comment]
Javac 22 seems to be producing bytecode with a load, even if the variable was unnamed.
A further improvement could be for SpotBugs to warn that a named variable _could_ be unnamed, is that what you're suggesting?

**Comment by boris-petrov** [comment]
Yes. I guess the `DLS_DEAD_LOCAL_STORE` check could be reused (because a named variable that can be unnamed means it's unused) or perhaps a new check could be implemented if that's going to be easier as it must be enabled only on JDK 22+.

**Comment by JuditKnoll** [comment]
As far as I can see this issue was fixed by https://github.com/spotbugs/spotbugs/pull/2937 .
